Migrate `pattern-export-to-repo` command

GeoJunkie opened the pull request:
This is the first part of #6 .

Migrates the `pattern-export-to-repo` command from the old CLI tool.

The command should not be launched until the image replacement updates are complete (https://github.com/a8cteam51/team51-cli/issues/6#issuecomment-2069454964), but this needs a review to make sure it fits with the new architecture.

GeoJunkie committed:
Copy pattern-export-to-repo command

This is a direct copy and needs to be refactored before it will work.

GeoJunkie committed:
Update command to match new structure

GeoJunkie committed:
WIP

GeoJunkie committed:
Command functionality restored

GeoJunkie committed:
Formatting/linting

GeoJunkie committed:
Fix so options work from the command line

GeoJunkie committed:
Rename branch to be identified as a pattern by GH workflow

GeoJunkie committed:
Add function to create a PR

GeoJunkie committed:
Revert "Add function to create a PR"

This reverts commit 564c94442a2574b6e7ec969e4c05773e845e9655.

This will be done later. Ops Oasis will need an update as well.

GeoJunkie committed:
Output GH details

GeoJunkie committed:
Link to the correct repo and directly to the diff

ahegyes commented:
I typically recommend the following naming structure:

* file: `<service>_<object>_<descriptors>_<action>`
* command: `<service>:<action>_<object>_<descriptors>` 

This keeps commands related to each other sorted in the IDE (all DeployHQ commands are "together", then all "DeployHQ project" commands, then all "DeployHQ project server" commands, and so on...). 

At the same time, it's more intuitive to type in a command that starts with a verb (you type in what you're doing, then press tab for options). 

Based on that, I would rename the file `GitHub_Pattern_To_Repo_Export` and the command to `github:export-pattern-to-repo`. But this is merely a suggestion ... If you think the current names make more sense for this particular case, cool 😄

ahegyes commented on commands/GitHub_Pattern_Export_To_Repo.php:
```diff
@@ -0,0 +1,265 @@
+		if ( $output->isVerbose() ) {
+			$output->writeln( "<info>Site {$this->pressable_site->id}: {$this->pressable_site->url}</info>" );
+		}
+		if ( \is_null( $this->pressable_site ) ) {
```
You don't need this check. `get_pressable_site_input` will auto-exit if it can't get the site. Functions that start with `maybe_*` will allow a value of `null` to return, so you'd need the check for `maybe_get_pressable_site_input` but not this one.

GeoJunkie replied:
TIL! Removed that check in 214ff98

ahegyes commented on commands/GitHub_Pattern_Export_To_Repo.php:
```diff
@@ -0,0 +1,265 @@
+		}
+
+		// Store the ID of the site in the argument field.
+		$input->setArgument( 'site', $this->pressable_site->id );
```
I found out that it's possible to set anything programmatically in an `$input` argument or option, so I typically store the whole object in case it's useful later. Something like `$input->setArgument( 'site', $this->pressable_site )`. I just find it more useful and the other commands do it like that.

GeoJunkie replied:
I'm curious about this one, actually. What's the purpose of storing that back in the argument and setting it as a class property?

In the rest of the class, we just refer back to `$this->pressable_site` and never get information from the arguments again. Should we just skip setting that argument or is there value to it?

ahegyes commented on commands/GitHub_Pattern_Export_To_Repo.php:
```diff
@@ -0,0 +1,265 @@
+		// Store the ID of the site in the argument field.
+		$input->setArgument( 'site', $this->pressable_site->id );
+
+		$this->pattern_name = $input->getArgument( 'pattern-name' );
```
Could you use `get_string_input` here? It does exactly this ... checks arguments, checks options, and if not there, prompts the user ... even better, it will exit the command with an error if it can't get it!

GeoJunkie replied:
Another TIL for me :) 

Changed in c67471f

ahegyes commented on commands/GitHub_Pattern_Export_To_Repo.php:
```diff
@@ -0,0 +1,265 @@
+		}
+
+		// Check if the category slug was already provided as an argument. If not, prompt the user for it.
+		$this->category_slug = $input->getArgument( 'category-slug' );
```
Ditto ... try using `get_string_input` here 😄

GeoJunkie replied:
That actually became redundant since `get_string_input()` added above took care of all that! 🤣

ahegyes commented on commands/GitHub_Pattern_Export_To_Repo.php:
```diff
@@ -0,0 +1,265 @@
+			$input->setArgument( 'category-slug', $this->category_slug );
+		}
+		if ( $output->isVerbose() ) {
+			$output->writeln( "<info>Category slug: {$this->category_slug}</info>" );
```
Instead of wrapping everything in `if ( $output->isVerbose() )`, you can simply set the verbosity on the `writeln` method:

`$output->writeln( "<info>Category slug: {$this->category_slug}</info>", OutputInterface::VERBOSITY_VERBOSE );`

That will only be outputted on commands called with `-v` or `-vv` or `-vvv`.

GeoJunkie replied:
Nice! Updated in 2692994

ahegyes commented on commands/GitHub_Pattern_Export_To_Repo.php:
```diff
@@ -0,0 +1,265 @@
+	 * {@inheritDoc}
+	 */
+	protected function execute( InputInterface $input, OutputInterface $output ): int {
+
```
I typically start the `execute` function by outputting a line that details what's being done with all the info possibly required to reproduce (in this case, the site info and so on). Example from another command:

```
$output->writeln( "<fg=magenta;options=bold>Opening an interactive $this->shell_type shell for {$this->site->displayName} (ID {$this->site->id}, URL {$this->site->url}) as $this->email.</>" );
```

ahegyes replied:
And yes, the `magenta` and `bold` are important because that's the only output that looks like it. So you know what to look for 😄

GeoJunkie replied:
Added that messaging in 42466e0. I also updated the color of the later success message so that initial message is the only output that looks like it ;)

ahegyes commented on commands/GitHub_Pattern_Export_To_Repo.php:
```diff
@@ -0,0 +1,265 @@
+	protected function execute( InputInterface $input, OutputInterface $output ): int {
+
+		// Upload script.
+		$sftp   = \Pressable_Connection_Helper::get_sftp_connection( $this->pressable_site->id );
```
This can fail more often that you'd think ... I'd add a `null` check on `$sftp`. 

Like this:

```
if ( \is_null( $ssh ) ) {
    $output->writeln( '<error>Could not connect to the SSH server.</error>' );
    return Command::FAILURE;
}
```

ahegyes replied:
Maybe also rename `$sftp` to `$sftp_connection` akin to `$ssh_connection` a few lines below?

GeoJunkie replied:
Added in f32bd4f

ahegyes commented on commands/GitHub_Pattern_Export_To_Repo.php:
```diff
@@ -0,0 +1,265 @@
+
+		// Run script.
+		$result = $ssh_connection->exec( sprintf( "wp eval-file /htdocs/pattern-extract.php '%s'", $this->pattern_name ) );
+		if ( $output->isDebug() ) {
```
Simplify this by using the `OutputInterface::VERBOSITY_DEBUG` option on `writeln` 😄

GeoJunkie replied:
Fixed in 2692994

ahegyes commented on commands/GitHub_Pattern_Export_To_Repo.php:
```diff
@@ -0,0 +1,265 @@
+			// Check if metadata.json exists before creating or overwriting
+			if ( ! file_exists( $metadata_path ) ) {
+				$metadata = array( 'title' => $this->category_slug );
+				file_put_contents( $metadata_path, json_encode( $metadata, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES ) );
```
Instead of `json_encode`, I recommend `encode_json_content` which comes with better error output if something goes wrong (hepler in `functions.php`).

ahegyes replied:
Ditto for replacing `json_decode` with `decode_json_content`.

GeoJunkie replied:
Those are nice! Switched the calls in 47a5ca6

ahegyes commented on commands/GitHub_Pattern_Export_To_Repo.php:
```diff
@@ -0,0 +1,265 @@
+	private function prompt_site_input( InputInterface $input, OutputInterface $output ): ?string {
+		// Ask for the pattern name, providing an example as a hint.
+		$question = new Question( '<question>Enter the site ID or URL to extract the pattern from:</question> ' );
+		$question->setAutocompleterValues( \array_map( static fn( object $site ) => $site->url, \get_pressable_sites() ?? array() ) );
```
Instead of `array_map`, we can simplify to `array_column`:

```
$question->setAutocompleterValues( \array_column( get_pressable_sites() ?? array(), 'url' ) );
```

GeoJunkie replied:
Updated in 673f32c

ahegyes commented on commands/GitHub_Pattern_Export_To_Repo.php:
```diff
@@ -0,0 +1,265 @@
+		$question->setAutocompleterValues( \array_map( static fn( object $site ) => $site->url, \get_pressable_sites() ?? array() ) );
+
+		// Retrieve the user's input.
+		$site = $this->getHelper( 'question' )->ask( $input, $output, $question );
```
You can simply do:

```
return $this->getHelper( 'question' )->ask( $input, $output, $question );
```

If there is nothing from the user, that will return `null` so same thing.

GeoJunkie replied:
Oh, interesting. Reduced that down in 425fd81

ahegyes replied:
Technically it will return whatever the `$default` parameter was on the `Question` object and that is `null` by default. Just a small technicality.

ahegyes commented on commands/GitHub_Pattern_Export_To_Repo.php:
```diff
@@ -0,0 +1,265 @@
+	 * @return string|null
+	 */
+	private function prompt_pattern_name_input( InputInterface $input, OutputInterface $output ): ?string {
+		if ( $input->isInteractive() ) {
```
No need to check for `$input->isInteractive()` actually. I've learned that, when the console is in non-interactive mode, the `Question` object simply returns the default value (if it has one set as the second argument which is `null` by default).

GeoJunkie replied:
I'd actually removed that at one point, and put it back when adding some other logic back in. I wasn't sure if Tiago had included it for a reason I missed, but since you've confirmed it's not needed I took it back out in be8f23a (and also adjusted the return similar to  https://github.com/a8cteam51/team51-cli/pull/29/files/d34e25cac18aa52dccbd6019079e2391097f4a74#r1580624188 ) :)

ahegyes replied:
I used it on the old CLI like that but I realized that it's not actually needed when I started rebuilding the new one. So it's probably on me because people followed the pattern I laid out there 😅

ahegyes commented on commands/GitHub_Pattern_Export_To_Repo.php:
```diff
@@ -0,0 +1,265 @@
+			$category_slug = $this->getHelper( 'question' )->ask( $input, $output, $question );
+
+			// Ensure the input matches the expected format.
+			$category_slug = $this->slugify( $category_slug );
```
Could `$this->slugify` be replaced with the `slugify` helper in `functions.php`? Also, would it make more sense to run `slugify` inside `initialize` after fetching the value?

GeoJunkie replied:
Makes sense. `slugify()` didn't exist on the old architecture, but now that it does, it makes sense to use the central one. The outcome of both functions is the same, so I removed the one in the class and pointed to the one in `functions.php`. Changed in 36742e8

GeoJunkie replied:
> Also, would it make more sense to run slugify inside initialize after fetching the value?

I restructured this in 3793653 to use `get_string_input()` and then slugify the results from that, so it always slugifies what comes in.

ahegyes commented on scaffold/pattern-extract.php:
```diff
@@ -0,0 +1,56 @@
+		'syncStatus' => '',
+	);
+
+	echo json_encode( $response );
```
Since this script will be running on WP, should we replace `json_encode` with `wp_json_encode`?

GeoJunkie replied:
Definitely. Updated in f13f1c6

ahegyes commented on commands/GitHub_Pattern_Export_To_Repo.php:
```diff
@@ -0,0 +1,265 @@
+		}
+
+		// Run script.
+		$result = $ssh_connection->exec( sprintf( "wp eval-file /htdocs/pattern-extract.php '%s'", $this->pattern_name ) );
```
I think this would be more robust like this:

```
sprintf( "wp eval-file /htdocs/pattern-extract.php %s", escapeshellarg( $this->pattern_name ) )
```

GeoJunkie replied:
That makes a lot of sense. Updated in 8053668

ahegyes commented:
After reviewing what the command itself does, I do have a philosophical suggestion -- what if we split the command in two? Technically, each command is supposed to "do one thing on one service". If we need multiple things done, then we should chain the commands.

That way, if anything fails, we can easily continue the process by calling the command that failed with the appropriate arguments. Look at the `pressable:site-create` command which will now make calls to `deployhq:create-project` and `deployhq:create-project-server`. This way, if creating the DeployHQ stuff fails for whatever reason, we only need to call those commands instead of doing everything manually. 

Moreover, maybe there will be a legitimate use-case of exporting the WP Pattern and doing something else with it, not just uploading it to GitHub. Or, maybe we want to reuse the `upload to GitHub` part on sites not hosted on Pressable. 

So I would suggest these two commands:

1) `pressable:export-site-pattern` hosted in `Pressable_Site_Pattern_Export.php` (later we can then create a command called `wpcom:export-site-pattern` hosted in `WPCOM_Site_Pattern_Export.php`)

2) `github:commit-site-pattern` hoted in `GitHub_Site_Pattern_Commit.php` 

Thoughts?

ahegyes commented on scaffold/pattern-extract.php:
```diff
@@ -0,0 +1,56 @@
+	}
+}
+
+// Check if $result is populated, otherwise set default values to prevent errors.
```
I don't think this comment is true anymore ... maybe it needs updating/deleting?

GeoJunkie replied:
Yep. Updated in ef0f75e to reflect what's happening here.

GeoJunkie committed:
Rename class and command

Ref: https://github.com/a8cteam51/team51-cli/pull/29\#issuecomment-2078805153

GeoJunkie commented:
> I typically recommend the following naming structure:
> 
> * file: `<service>_<object>_<descriptors>_<action>`
> * command: `<service>:<action>_<object>_<descriptors>`
> 
> This keeps commands related to each other sorted in the IDE (all DeployHQ commands are "together", then all "DeployHQ project" commands, then all "DeployHQ project server" commands, and so on...).
> 
> At the same time, it's more intuitive to type in a command that starts with a verb (you type in what you're doing, then press tab for options).
> 
> Based on that, I would rename the file `GitHub_Pattern_To_Repo_Export` and the command to `github:export-pattern-to-repo`. But this is merely a suggestion ... If you think the current names make more sense for this particular case, cool 😄

I was basing this on the old setup where the command name and the filename matched. This makes sense, so I updated it :). (Commit d91df5b )

GeoJunkie commented:
> After reviewing what the command itself does, I do have a philosophical suggestion -- what if we split the command in two? Technically, each command is supposed to "do one thing on one service". If we need multiple things done, then we should chain the commands.
> 
> That way, if anything fails, we can easily continue the process by calling the command that failed with the appropriate arguments. Look at the `pressable:site-create` command which will now make calls to `deployhq:create-project` and `deployhq:create-project-server`. This way, if creating the DeployHQ stuff fails for whatever reason, we only need to call those commands instead of doing everything manually.
> 
> Moreover, maybe there will be a legitimate use-case of exporting the WP Pattern and doing something else with it, not just uploading it to GitHub. Or, maybe we want to reuse the `upload to GitHub` part on sites not hosted on Pressable.
> 
> So I would suggest these two commands:
> 
> 1. `pressable:export-site-pattern` hosted in `Pressable_Site_Pattern_Export.php` (later we can then create a command called `wpcom:export-site-pattern` hosted in `WPCOM_Site_Pattern_Export.php`)
> 2. `github:commit-site-pattern` hoted in `GitHub_Site_Pattern_Commit.php`
> 
> Thoughts?

@ahegyes I agree in part. I was thinking about splitting it up while I was migrating it but had initially decided against it. Give the "do one thing on one service" philosophy, it makes sense to have the functions of the command split up, but I also think having one command to carry out all the steps in the most common process makes a lot of sense here, too. We'll be asking TAMs to do this task a lot, and if they can do it in one step instead of two that would be better.

Can we do something similar to `pressable:site-create` where one service command calls the others?

I'm thinking we would have `pressable:export-pattern-to-library` (with the potential to add `wpcom:export-pattern-to-library` later). This command would then chain the two commands: `pressable:export-site-pattern` and `github:add-pattern`, and pass the exported file from one to the other.

I renamed the commands slightly to cover what they would do:
- `pressable:export-pattern-to-library` takes a pattern from a site and exports it to the pattern library.
-  `pressable:export-site-pattern` - exports a pattern from a site.
- `github:add-pattern` - adds the pattern to the library. At this point, the pattern shouldn't be affiliated with a particular site. (In a future iteration, I'm thinking we may also add the PR creation to this command instead of relying on the [GitHub action](https://github.com/a8cteam51/team51-patterns/blob/trunk/.github/workflows/pr-on-new-branch.yml). This would allow us to give them a PR link instead of just the branch.)

What do you think?

GeoJunkie committed:
Remove redundant check for invalid site

Ref: https://github.com/a8cteam51/team51-cli/pull/29/files/d34e25cac18aa52dccbd6019079e2391097f4a74\#r1580609083

GeoJunkie committed:
Use get_string_input()

Ref: https://github.com/a8cteam51/team51-cli/pull/29/files/d34e25cac18aa52dccbd6019079e2391097f4a74\#r1580611951

GeoJunkie committed:
Simplify verbose and debug messaging calls

Ref:
https://github.com/a8cteam51/team51-cli/pull/29/files/d34e25cac18aa52dccbd6019079e2391097f4a74\#r1580614018

GeoJunkie committed:
Add message detailing task to carry out

Also updated coloring of success message
Ref: https://github.com/a8cteam51/team51-cli/pull/29/files/d34e25cac18aa52dccbd6019079e2391097f4a74\#r1580615342

GeoJunkie committed:
Add check for SFTP connection

Ref: https://github.com/a8cteam51/team51-cli/pull/29/files/d34e25cac18aa52dccbd6019079e2391097f4a74\#r1580617158

GeoJunkie committed:
Escape shell argument

Ref: https://github.com/a8cteam51/team51-cli/pull/29/files/d34e25cac18aa52dccbd6019079e2391097f4a74\#r1580634529

GeoJunkie committed:
Use encode_json_content() and encode_json_content()

Ref: https://github.com/a8cteam51/team51-cli/pull/29/files/d34e25cac18aa52dccbd6019079e2391097f4a74\#r1580621144

GeoJunkie committed:
Use array_column() instead of array_map()

Ref: https://github.com/a8cteam51/team51-cli/pull/29/files/d34e25cac18aa52dccbd6019079e2391097f4a74\#r1580623342

GeoJunkie committed:
Directly return question results

Ref: https://github.com/a8cteam51/team51-cli/pull/29/files/d34e25cac18aa52dccbd6019079e2391097f4a74\#r1580624188

GeoJunkie committed:
Removed unneccessary isInteractive check

Also directly returned data from the question

Ref: https://github.com/a8cteam51/team51-cli/pull/29/files/d34e25cac18aa52dccbd6019079e2391097f4a74\#r1580625226

GeoJunkie committed:
Use slugify() from functions.php

Ref: https://github.com/a8cteam51/team51-cli/pull/29/files/d34e25cac18aa52dccbd6019079e2391097f4a74\#r1580626607

GeoJunkie committed:
For category slug, use get_string_input() and slugify the value always

Ref: https://github.com/a8cteam51/team51-cli/pull/29/files/d34e25cac18aa52dccbd6019079e2391097f4a74\#r1580626607

GeoJunkie committed:
Update comment to reflect what it does now

Ref: https://github.com/a8cteam51/team51-cli/pull/29/files/d34e25cac18aa52dccbd6019079e2391097f4a74\#r1580655625

GeoJunkie committed:
Use wp_json_encode() in the on-site script

GeoJunkie reviewed (commented):
Thanks @ahegyes  for the thorough review. I corrected most things, although I had a couple small questions for you.

ahegyes commented:
> What's the purpose of storing that back in the argument and setting it as a class property?

If the argument is marked as mandatory, Symfony will exit the command if it's not set. We can set it programmatically in the `initialize` method so that's what we do. This is not required if the argument is optional but I like to do it always because it keeps the app status consistent.

As for keeping it a class property ... well, you're right, we could technically fetch it always from the input arguments. I _think_ that I actually do that in 1-2 commands when I just needed it for an input helper ... but it's more prone to typos and it doesn't have IDE auto-complete features. 

I hadn't given this too much though beyond always setting it on `$input` to keep Symfony happy 😅 Maybe we should reuse that more often ...

ahegyes commented:
> Can we do something similar to pressable:site-create where one service command calls the others?

Absolutely, and that's what I meant! We could add a negatable option (`InputOption::VALUE_NEGATABLE`, see https://symfony.com/doc/current/console/input.html) like `--commit-to-github` which would then chain the commands (or not ... not sure whether the default should be `yes` or `no`). 

> I renamed the commands slightly to cover what they would do: [...] What do you think?

I don't think we really need 3 commands. We can just do this:

1. `github:commit-pattern` (or `github:commit-gutenberg-pattern` or `github:commit-gb-pattern` or `github:commit-wp-pattern`, really not sure what my favorite is ... thoughts?) 
2. `pressable:export-site-pattern` with the negatable option `--commit-to-github` in order to chain or not chain the GH command; as said above, not sure if `--commit` or `--no-commit` should be default

ahegyes commented on commands/GitHub_Pattern_To_Repo_Export.php:
```diff
@@ -0,0 +1,226 @@
+/**
+ * Exports a block pattern from a site to the block pattern library.
+ */
+#[AsCommand( name: 'github:export-pattern-to-repo	' )]
```
Does this actually work with the empty space at the end? Symfony must be trimming the command names somewhere.

GeoJunkie replied:
It does! Thankfully, Symfony is smarter than I am, but still I removed the extra space in 1a10c76

ahegyes commented on commands/GitHub_Pattern_To_Repo_Export.php:
```diff
@@ -0,0 +1,226 @@
+			->setHelp( 'This command exports a specified block pattern into a category within a GitHub repository.' )
+			->addArgument( 'site', InputArgument::REQUIRED, 'ID or URL of the Pressable site to run the command on.' )
+			->addArgument( 'pattern-name', InputArgument::REQUIRED, 'The unique identifier of the block pattern to export (e.g., "namespace/pattern-name").' )
+			->addArgument( 'category-slug', InputArgument::REQUIRED, 'The slug of the category under which the pattern should be exported. It should be lowercase with hyphens instead of spaces (e.g., "featured-patterns").' );
```
Very small nitpick but I tried to use ` instead of ' or " in descriptions throughout the CLI since that one never needs to be escaped. It would make sense for consistency to use it here to for `featured-patterns`.

GeoJunkie replied:
I fixed that in 38a55d9

I didn't use backticks because it doesn't render `like this` on the command line, but if that's the standard within the CLI, it makes sense to be consistent for readability and ease of use.

ahegyes commented on commands/GitHub_Pattern_To_Repo_Export.php:
```diff
@@ -0,0 +1,226 @@
+
+		// Retrieve the given site.
+		$this->pressable_site = get_pressable_site_input( $input, $output, fn() => $this->prompt_site_input( $input, $output ) );
+		$output->writeln( "<info>Site {$this->pressable_site->id}: {$this->pressable_site->url}</info>", Output::VERBOSITY_VERBOSE );
```
The `<info>` output is green and can be easily confused with the `success` one, in my opinion. So I typically used `<comment>` when something informative was outputted. It might make sense to do that, for consistency with the other commands.

GeoJunkie replied:
Fixed in a21e44d

GeoJunkie replied:
Given that new information about the `<info>` output, I altered the `Done!` message at the end to use that tag :)

ahegyes commented on commands/GitHub_Pattern_To_Repo_Export.php:
```diff
@@ -0,0 +1,226 @@
+	 */
+	protected function execute( InputInterface $input, OutputInterface $output ): int {
+
+		$output->writeln( "<fg=magenta;options=bold>Exporting {$this->pattern_name} (Category: {$this->category_slug}) from {$this->pressable_site->displayName} (ID {$this->pressable_site->id}, URL {$this->pressable_site->url})</>" );
```
Since we have this informative output as the very first thing being outputted, do you think we still need the `<info>` statements from the `initialize` method? 🤔

GeoJunkie replied:
The (now) `<comment>` in the `initialize` method only appears when running the CLI in Verbose mode. It's useful for troubleshooting if something odd is happening.

ahegyes commented on commands/GitHub_Pattern_To_Repo_Export.php:
```diff
@@ -0,0 +1,226 @@
+		}
+		$result = $sftp_connection->put( '/htdocs/pattern-extract.php', file_get_contents( __DIR__ . '/../scaffold/pattern-extract.php' ) );
+		if ( ! $result ) {
+			$output->writeln( "<error>Failed to copy pattern-extract.php to {$this->pressable_site->id}.</error>" );
```
If this fails, we can't really do the export. Shouldn't we return a `Command::FAILURE`?

ahegyes replied:
We could do fewer checks here actually. Since the CLI is running on PHP8+, we don't have to check for `is_null( $sftp_connection )`. We can just:

```
$result = $sftp_connection?->put( ... );
if ( empty( $result ) ) {
   // return
}
```

That `?` will make sure that `put` only gets called if `$sftp_connection` is not `null`. If it is, then `$result` will contain `null`. So we can get away with just one check.

ahegyes replied:
Ditto down below with `$ssh_connection?->exec`.

GeoJunkie replied:
> If this fails, we can't really do the export. Shouldn't we return a Command::FAILURE?

Yes. Added that in 381d6c0

> That ? will make sure that put only gets called if $sftp_connection is not null. If it is, then $result will contain null. So we can get away with just one check.

I guess this depends on how much information we want to give when something goes wrong. With the separate checks, you'll know whether you are having a problem connecting to the site, or if there's a problem with actually pushing the file onto the server/running the command.

If you want to keep it standardized and do the single check I'll make the update, but maybe we should keep it that way for granularity when something goes wrong?

ahegyes replied:
We should probably move the check for the connection inside the Connection_Helper class actually. That way, we always get the info in the terminal.

But we can leave things as-is for now.

GeoJunkie replied:
Looking [further down](https://github.com/a8cteam51/team51-cli/blob/381d6c0253a65dece626229998db07b24efa0b71/commands/GitHub_Pattern_To_Repo_Export.php#L105-L106) I think this is a carry-over when I copy/pasted the code. The file isn't currently deleted from the server because the line is commented out:

```php
		// Delete script.
		// $ssh_connection->exec( 'rm /htdocs/pattern-extract.php' );
```

So if the put fails, it's because the file is already there. This made sense to save time while building it but I think that's not a great assumption to make now, so I return a failure in  381d6c0 and uncommented the `rm` command in 508f656

GeoJunkie replied:
> We should probably move the check for the connection inside the Connection_Helper class actually. That way, we always get the info in the terminal.

That's a good idea. We could consider returning details, as well, if it would be helpful to give that to the calling function.

ahegyes commented on commands/GitHub_Pattern_To_Repo_Export.php:
```diff
@@ -0,0 +1,226 @@
+		if ( ! empty( $result ) ) {
+
+			// Temporary directory to clone the repository
+			$temp_dir = sys_get_temp_dir() . '/team51-patterns';
```
Do you think we should use a random folder name? If something goes wrong here, the `team51-patterns` directory won't be deleted. And then the command will fail in the future because the folder already exists. Maybe we just append a randomly generated string at the end? 

Like:

```
$temp_dir = sys_get_temp_dir() . DIRECTORY_SEPARATOR . uniqid( 'team51-patterns_', true );
```

GeoJunkie replied:
That's a good idea. Added in a81da8f

ahegyes committed:
Enhance README

GeoJunkie commented:
> > What's the purpose of storing that back in the argument and setting it as a class property?
> 
> If the argument is marked as mandatory, Symfony will exit the command if it's not set. We can set it programmatically in the `initialize` method so that's what we do. This is not required if the argument is optional but I like to do it always because it keeps the app status consistent.
> 
> As for keeping it a class property ... well, you're right, we could technically fetch it always from the input arguments. I _think_ that I actually do that in 1-2 commands when I just needed it for an input helper ... but it's more prone to typos and it doesn't have IDE auto-complete features.
> 
> I hadn't given this too much though beyond always setting it on `$input` to keep Symfony happy 😅 Maybe we should reuse that more often ...

This all makes sense! For now, I'll leave it as-is. As things update we can use it more :D

GeoJunkie committed:
Remove extra whitespace from command string

GeoJunkie committed:
Use backtick for values in help text

GeoJunkie committed:
Use `<comment>` instead of `<info>`

GeoJunkie committed:
Fail command if we can't upload the script

GeoJunkie committed:
Delete the script from the server

GeoJunkie committed:
Give a green info message on success

GeoJunkie committed:
Randomize tmp folder name

ahegyes commented on commands/GitHub_Pattern_To_Repo_Export.php:
```diff
@@ -0,0 +1,227 @@
+
+		// Check if the pattern name was already provided as an argument. If not, prompt the user for it.
+		$this->pattern_name = get_string_input( $input, $output, 'pattern-name', fn() => $this->prompt_pattern_name_input( $input, $output ));
+		$output->writeln( "<info>Pattern name: {$this->pattern_name}</info>", Output::VERBOSITY_VERBOSE );
```
Should we also update this to `<comment>` same as we did in line 62? Same for line 73? And since we're doing this for debugging purposes, what do you think about changing the verbosity to `VERBOSITY_DEBUG`? The current one is fine though.

GeoJunkie replied:
Good idea. Fixed it there and a couple other spots in 8560a3d

ahegyes commented on scaffold/pattern-extract.php:
```diff
@@ -0,0 +1,56 @@
+$pattern_registry = WP_Block_Patterns_Registry::get_instance();
+$pattern          = $pattern_registry->get_registered( $pattern_name );
+
+$result = array();
```
This is extremely nit-picky but what if `$result` was initially `null`? And we changed the checks down to `is_null` instead? Currently, I thought that `$result` will be populated with multiple entries but that's not the case -- we are using the empty array as a null-ish value which works great but I had to look at the code twice because I thought it's some sort of mistake in assignment on lines 24 and 39.

GeoJunkie replied:
I could see that getting confusing. `empty()` actually acts the same way with an empty array or a null value, but I updated the check to `is_null()` as well to help avoid that kind of confusion.

Updated in f8342d3

ahegyes reviewed (approved):
This look good to me, assuming that it works (I haven't tested it yet). I added a few comments but they can be ignored.

GeoJunkie committed:
Use <comment> vs. <info>

GeoJunkie committed:
Assign null instead of empty array for readability
